<a href="https://colab.research.google.com/github/hasonsk/Finetune-LLM-for-deteting-cookie/blob/main/Qwen2_5_3B_Instruct_SFT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

Read our **[Qwen3 Guide](https://docs.unsloth.ai/basics/qwen3-how-to-run-and-fine-tune)** and check out our new **[Dynamic 2.0](https://docs.unsloth.ai/basics/unsloth-dynamic-2.0-ggufs)** quants which outperforms other quantization methods!

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl==0.15.2 triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf datasets huggingface_hub hf_transfer
    !pip install --no-deps unsloth
    !pip install -U datasets

In [2]:
import json
import ast
import pandas as pd
import numpy as np
import torch
from datasets import Dataset, load_dataset
# from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
from tqdm import tqdm
from typing import List, Dict, Any
# from peft import PeftModel, LoraConfig, get_peft_model

### Unsloth

In [3]:
import os, wandb
from huggingface_hub import login

os.environ["WANDB_API_KEY"] = "810a489e4029d94651a3868c08a49812cc1e4cc7"
os.environ["HUGGINGFACE_API_KEY"] = "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"

In [4]:
hf_token = os.environ["HUGGINGFACE_API_KEY"]
login(hf_token)

secret_key = os.environ["WANDB_API_KEY"]
wandb.login(key=secret_key)

wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: sonhh2003 (sonhh2003-hanoi-university-of-science-and-technology) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


True

In [5]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 4096 # Choose any! We auto support RoPE Scaling internally!
dtype = None # None for auto detection. Float16 for Tesla T4, V100, Bfloat16 for Ampere+
load_in_4bit = True # Use 4bit quantization to reduce memory usage. Can be False.

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "Qwen/Qwen2.5-3B-Instruct",
    max_seq_length = max_seq_length,
    dtype = dtype,
    load_in_4bit=load_in_4bit,
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.5.1: Fast Qwen2 patching. Transformers: 4.51.3.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


We now add LoRA adapters so we only need to update 1 to 10% of all parameters!

In [6]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 16, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 16,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

Unsloth 2025.5.1 patched 36 layers with 36 QKV layers, 36 O layers and 36 MLP layers.


<a name="Data"></a>
### Data Prep
We now use the Alpaca dataset from [yahma](https://huggingface.co/datasets/yahma/alpaca-cleaned), which is a filtered version of 52K of the original [Alpaca dataset](https://crfm.stanford.edu/2023/03/13/alpaca.html). You can replace this code section with your own data prep.

**[NOTE]** To train only on completions (ignoring the user's input) read TRL's docs [here](https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only).

**[NOTE]** Remember to add the **EOS_TOKEN** to the tokenized output!! Otherwise you'll get infinite generations!

If you want to use the `llama-3` template for ShareGPT datasets, try our conversational [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Alpaca.ipynb)

For text completions like novel writing, try this [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Mistral_(7B)-Text_Completion.ipynb).

In [7]:
alpaca_prompt = """You are an expert AI assistant specialized in analyzing website cookie policies. Your task is to read the provided cookie policy text and extract detailed, structured information about each cookie mentioned, adhering strictly to the guidelines and format specified below.
RESPONSE Format: JSON following structure:

{{
  "is_specific": 0,
  "cookies": [
     {{
      "cookie_name": "cookie_name",
      "declared_purpose": "declared_purpose",
      "declared_retention": "declared_retention",
      "declared_third_parties": ["declared_third_parties"],
      "declared_description": "declared_description"
    }}, ...
  ]
}}

If no cookies are specifically described, only response {{"is_specific": 0, "cookies": []}}

### Instruction:
{}

### Input:
{}

### Response:
{{
  "is_specific": {},
  "cookies": {}
}}"""

SYSTEM_PROMPT = """
Extraction guidelines:
For "is_specific": Determines if the website provides detailed descriptions of individual cookies:
- Set to 0 if cookies are only described generically (e.g., "performance cookies," "necessary cookies," "Google cookies") without specific explanations
- Set to 1 if cookies are described with specific names, purposes, retention periods, and third parties. Example: "The '_ga' cookie is used by Google Analytics to distinguish users and is stored for 2 years" would result in is_specific = 1

For the "cookies" list containing objects, each object has
- "cookie_name": Extract the exact technical name as mentioned. One object just has only one cookie name, ìf more one, create multiple objects. For example: ga, _gid, _gat --> 3 object cookies
- "declared_purpose": cookie's purpose. With the following options:
  * Strictly Necessary: Essential for basic website functionality
  * Functionality: Personalizes user experience
  * Analytical: Collects usage data
  * Targeting/Advertising/Marketing: For personalized ads
  * Performance: Improves technical performance
  * Social Sharing: Enables social media integration
  * Null: When no specific purpose information is provided
- "declared_retention": Record the exact storage duration as mentioned
- "declared_third_parties": An list of third parties involved in the use of this cookie for website-owned cookies
- "declared_description": exact wording from the policy without modification or embellishment

Extract detailed information about each cookie mentioned in that policy"""


EOS_TOKEN = tokenizer.eos_token # Must add EOS_TOKEN
def formatting_prompts_func(example):
    if example['english_content']:
        content = f"Cookie policy: {example['english_content']}\nTable: {example['english_table']}"
    elif example['english_table']:
        content = f"Table: {example['english_table']}"
    else:
        content = f"Content: {example['english_table']}"

    instruction = SYSTEM_PROMPT
    input       = content
    is_specific = example['is_specific']
    cookies     = ast.literal_eval(example['extracted_cookies'])

    text = alpaca_prompt.format(instruction, input, is_specific, cookies) + EOS_TOKEN
    return { "text" : text, }

from datasets import load_dataset
dataset = load_dataset("sonhask/Extract_cookies_from_cookie_policy_specific", split = "train")
dataset = dataset.map(formatting_prompts_func)

README.md:   0%|          | 0.00/826 [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/34.5M [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/11.3M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/8808 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/2937 [00:00<?, ? examples/s]

Map:   0%|          | 0/8808 [00:00<?, ? examples/s]

In [8]:
dataset

Dataset({
    features: ['url', 'lang', 'original_text', 'tables', 'translated_tables', 'processed_text', 'english_content', 'english_table', 'label', 'is_specific', 'extracted_cookies', '__index_level_0__', 'text'],
    num_rows: 8808
})

In [ ]:
import json
import ast

def process_cookies(extracted_cookies):
    if isinstance(extracted_cookies, list):
        return extracted_cookies

    elif isinstance(extracted_cookies, str):
        try:
            return json.loads(extracted_cookies)
        except json.JSONDecodeError:
            try:
                return ast.literal_eval(extracted_cookies)
            except (SyntaxError, ValueError):
                print(f"Không thể phân tích dữ liệu cookies: {extracted_cookies[:100]}...")
                return []

    else:
        print(f"Loại dữ liệu không hỗ trợ: {type(extracted_cookies)}")
        return []

specific_cookie_count = 0

for entry in dataset:
    if entry['is_specific'] == 1:
        try:
            cookies_list = process_cookies(entry.get('extracted_cookies', []))

            cookies_count = len(cookies_list)
            specific_cookie_count += cookies_count

            if cookies_count > 0:
                print(f"Đã tìm thấy {cookies_count} cookies")
        except Exception as e:
            print(f"Lỗi không xác định: {str(e)}")
            print(f"Giá trị: {entry.get('extracted_cookies', 'N/A')[:100]}...")
            continue

print(f"Tổng số cookies khi is_specific là 1: {specific_cookie_count}")

In [10]:
def thong_ke(dataset):
    empty_label_count = 0
    non_empty_label_count = 0

    for i in dataset:
        if i['label'] == "[]":
            empty_label_count += 1
        else:
            non_empty_label_count += 1

    print("#[]: ", empty_label_count)
    print("#Has cookies: ", non_empty_label_count)

thong_ke(dataset)

#[]:  6471
#Has cookies:  2337


In [11]:
from datasets import concatenate_datasets

has_no_cookie = "[]"

empty_label_dataset = dataset.filter(lambda example: example['label'] == has_no_cookie)
non_empty_label_dataset = dataset.filter(lambda example: example['label'] != has_no_cookie)

# Downsample the non-empty dataset to the size of the empty dataset
n_samples = len(non_empty_label_dataset)
print("N samples: ", n_samples)
downsampled_empty_dataset = empty_label_dataset.shuffle(seed=42).select(range(n_samples))

# Gộp 2 dataset
combined_dataset = concatenate_datasets([non_empty_label_dataset, downsampled_empty_dataset])

# Shuffle dataset
balanced_dataset = combined_dataset.shuffle(seed=42)

# Verify the balanced dataset statistics
thong_ke(balanced_dataset)

Filter:   0%|          | 0/8808 [00:00<?, ? examples/s]

Filter:   0%|          | 0/8808 [00:00<?, ? examples/s]

N samples:  2337
#[]:  2337
#Has cookies:  2337


In [12]:
def tokenize_batch(examples):
    toks = tokenizer(
        examples["text"],
        padding="max_length",
        truncation=True,
        max_length=max_seq_length,
    )
    # Copy input_ids as labels for causal LM
    toks["labels"] = toks["input_ids"].copy()
    return toks

dataset = balanced_dataset.map(tokenize_batch, batched=True)

Map:   0%|          | 0/4674 [00:00<?, ? examples/s]

In [13]:
dataset = dataset.remove_columns(['url', 'lang', 'original_text', 'tables', 'processed_text', 'english_content', 'english_table', 'translated_tables', 'label', '__index_level_0__', 'text', ])
dataset = dataset.remove_columns(['is_specific', 'extracted_cookies'])

In [14]:
dataset

Dataset({
    features: ['input_ids', 'attention_mask', 'labels'],
    num_rows: 4674
})

In [15]:
run = wandb.init(
    project="Fine-tuning Qwen_2_5_Instruct_SFT (v3)",
    id="we4w5817",
    resume="must"
)

In [16]:
from trl import SFTTrainer, SFTConfig

args = SFTConfig(
    per_device_train_batch_size = 8,
    gradient_accumulation_steps = 4,
    warmup_steps = 5,
    num_train_epochs = 2,    # Set this for 1 full training run.
    learning_rate = 0.7001,
    logging_steps = 1,
    optim = "paged_adamw_32bit",
    weight_decay = 0.01,
    lr_scheduler_type = "cosine",
    seed = 3407,
    output_dir = "/content/drive/MyDrive/Project/Models-v3/Qwen2.5-3B-Instruct_SFT",
    report_to = "wandb",      # Use this for WandB etc
    save_strategy="steps",    # Hoặc "epoch" nếu bạn muốn lưu sau mỗi epoch
    save_steps=15,            # Lưu mỗi 50 steps (tùy bạn chỉnh)
    save_total_limit=3,       # Giới hạn số lượng checkpoint (xóa cũ)
    max_seq_length = max_seq_length,
    # max_steps = 60,
    label_names=["labels"],
    packing = False
)

trainer = SFTTrainer(
    model = model,
    processing_class = tokenizer,
    train_dataset = dataset,
    args=args,
)

In [17]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
3.281 GB of memory reserved.


In [ ]:
trainer_stats = trainer.train(resume_from_checkpoint=True)

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 4,674 | Num Epochs = 2 | Total steps = 292
O^O/ \_/ \    Batch size per device = 8 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (8 x 4 x 1) = 32
 "-____-"     Trainable parameters = 29,933,568/3,000,000,000 (1.00% trained)
	save_steps: 15 (from args) != 25 (from trainer_state.json)
wandb: WARNING The `run_name` is currently set to the same value as `TrainingArguments.output_dir`. If this was not intended, please specify a different run name by setting the `TrainingArguments.run_name` parameter.


Unsloth: Will smartly offload gradients to save VRAM!


In [ ]:
output_dir = "/content/drive/MyDrive/Project/Models-v3/Qwen2.5-3B-Instruct_SFT/saved-model-epoch2"
model.save_pretrained(output_dir)  # Local saving
tokenizer.save_pretrained(output_dir)

# output_to_hub = "sonhask/Meta-Llama-3.1-8B-Instruct-bnb-4bit_Extract_cookies"
# model.push_to_hub(output_to_hub) # Online saving
# tokenizer.push_to_hub(output_to_hub) # Online saving

In [ ]:
run.finish()

In [ ]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

<a name="Inference"></a>
### Inference
Let's run the model! You can change the instruction and input - leave the output blank!

**[NEW] Try 2x faster inference in a free Colab for Llama-3.1 8b Instruct [here](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Unsloth_Studio.ipynb)**

In [ ]:
from transformers import TextStreamer

# alpaca_prompt = Copied from above
FastLanguageModel.for_inference(model) # Enable native 2x faster inference

alpaca_prompt = """Below is an instruction that describes a task, paired with an input that provides further context.
Your task is read a cookie policy text and extract detailed information about each cookie mentioned in that policy.
RESPONSE Format: JSON following structure:

{{
  "is_specific": 0,
  "cookies": [
     {{
      "cookie_name": "cookie_name",
      "declared_purpose": "declared_purpose",
      "declared_retention": "declared_retention",
      "declared_third_parties": ["declared_third_parties"],
      "declared_description": "declared_description"
    }}, ...
  ]
}}
If no cookies are specifically described, only response {{"is_specific": 0, "cookies": []}}

### Instruction:
{}

### Input:
{}

### Response:
{{
  "is_specific": {},
  "cookies": {}
}}"""

SYSTEM_PROMPT = """
Extraction guidelines:
For "is_specific": Determines if the website provides detailed descriptions of individual cookies:
- Set to 0 if cookies are only described generically (e.g., "performance cookies," "necessary cookies," "Google cookies") without specific explanations
- Set to 1 if cookies are described with specific names, purposes, retention periods, and third parties. Example: "The '_ga' cookie is used by Google Analytics to distinguish users and is stored for 2 years" would result in is_specific = 1

For the "cookies" list containing objects, each object has
- "cookie_name": Extract the exact technical name as mentioned. One object just has only one cookie name, ìf more one, create multiple objects. For example: ga, _gid, _gat --> 3 object cookies
- "declared_purpose": cookie's purpose. With the following options:
  * Strictly Necessary: Essential for basic website functionality
  * Functionality: Personalizes user experience
  * Analytical: Collects usage data
  * Targeting/Advertising/Marketing: For personalized ads
  * Performance: Improves technical performance
  * Social Sharing: Enables social media integration
  * Null: When no specific purpose information is provided
- "declared_retention": Record the exact storage duration as mentioned
- "declared_third_parties": An list of third parties involved in the use of this cookie for website-owned cookies
- "declared_description": exact wording from the policy without modification or embellishment"""


EOS_TOKEN = tokenizer.eos_token # Must add EOS_TOKEN
def formatting_prompts_func(example):
    if example['english_content']:
        content = f"Cookie policy: {example['english_content']}\nTable: {example['english_table']}"
    elif example['english_table']:
        content = f"Table: {example['english_table']}"
    else:
        content = f"Content: {example['english_table']}"

    instruction = SYSTEM_PROMPT
    input       = content
    is_specific = example['is_specific']
    cookies     = ast.literal_eval(example['extracted_cookies'])

    text = alpaca_prompt.format(instruction, input, is_specific, cookies) + EOS_TOKEN
    return { "text" : text, }

In [ ]:
# testdata =
balanced_dataset['text'][0]

In [ ]:
# def infer(example):
#     prompt = formatting_prompts_func(example)
#     inputs = tokenizer(
#         prompt,
#         return_tensors="pt",
#     ).to("cuda")

#     text_streamer = TextStreamer(tokenizer)
#     output = model.generate(**inputs, streamer=text_streamer, max_new_tokens=128)

def extract_response(_text):
    result_text = _text.split("### Response:\n")[-1].strip()
    # # Trích xuất phần JSON
    json_start = result_text.find("{")
    json_end = result_text.rfind("}")
    if json_start != -1 and json_end != -1 and json_end > json_start:
        json_text = result_text[json_start:json_end+1]
        return json_text

    return result_text

def generate_response(messages, max_new_tokens=4096, temperature=0):

    inputs = tokenizer(messages, return_tensors="pt", truncation=True,
                       max_length=max_new_tokens
                       ).to("cuda")

    # Tạo đầu ra
    with torch.no_grad():
        outputs = model.generate(
            **inputs,
            # max_new_tokens=max_new_tokens,
            do_sample=temperature > 0,
            temperature=temperature,
            top_p=1,
            pad_token_id=tokenizer.eos_token_id
        )

    # # Decode the prediction
    decoded_output = tokenizer.decode(outputs[0], skip_special_tokens=True)
    # result_text = decoded_output.split("### Response:\n")[-1].strip()


    # # # Trích xuất phần JSON
    # json_start = result_text.find("{")
    # json_end = result_text.rfind("}")
    # if json_start != -1 and json_end != -1 and json_end > json_start:
    #     json_text = result_text[json_start:json_end+1]
    #     return json_text

    # return result_text
    return extract_response(decoded_output)

In [ ]:
import json
import ast
import jsonschema
import numpy as np
from typing import List, Dict, Any, Union, Tuple
from collections import defaultdict
from sentence_transformers import SentenceTransformer
import sklearn.metrics.pairwise as pairwise
import re

# Schema cho dữ liệu cookie đã cập nhật
COOKIE_SCHEMA = {
    "type": "object",
    "required": ["is_specific", "cookies"],
    "properties": {
        "is_specific": {
            "type": "integer",
            "enum": [0, 1],
        },
        "cookies": {
            "type": "array",
            "items": {
                "type": "object",
                "required": [
                    "cookie_name",
                    "declared_purpose",
                    "declared_retention",
                    "declared_third_parties",
                    "declared_description"
                ],
                "properties": {
                    "cookie_name": {
                        "type": "string",
                    },
                    "declared_purpose": {
                        "type": "string",
                    },
                    "declared_retention": {
                        "type": ["string", "null"],
                    },
                    "declared_third_parties": {
                        "type": "array",
                        "items": {
                            "type": "string"
                        },
                    },
                    "declared_description": {
                        "type": ["string", "null"],
                    }
                }
            }
        }
    }
}

# 1. JSON Validity
def validate_json_format(predicted_json_str: str) -> bool:
    """
    Kiểm tra xem chuỗi JSON có hợp lệ theo schema mới hay không

    Args:
        predicted_json_str: Chuỗi JSON cần kiểm tra

    Returns:
        bool: True nếu JSON hợp lệ, False nếu không
    """
    try:
        # json_data = json.loads(predicted_json_str)
        json_data = ast.literal_eval(predicted_json_str)
        print(type(json_data))
        jsonschema.validate(instance=json_data, schema=COOKIE_SCHEMA)
        return True
    except (json.JSONDecodeError, jsonschema.exceptions.ValidationError):
        return False


In [ ]:
i = int(input())
print("---> Actual: ", extract_response(balanced_dataset['text'][i]))
predict = generate_response(balanced_dataset['text'][i])
print("---> Predict: ", predict)
extract_response(balanced_dataset['text'][i]) == predict

 You can also use a `TextStreamer` for continuous inference - so you can see the generation token by token, instead of waiting the whole time!

In [ ]:
def extract_valid_cookies(json_strings: List[str]) -> List[Dict]:
    """
    Trích xuất dữ liệu từ các chuỗi JSON và chỉ giữ lại các JSON hợp lệ

    Args:
        json_strings: Danh sách các chuỗi JSON

    Returns:
        List[Dict]: Danh sách các đối tượng JSON hợp lệ
    """
    result = []
    for js in json_strings:
        try:
            data = json.loads(js)
            jsonschema.validate(instance=data, schema=COOKIE_SCHEMA)
            result.append(data)
        except (json.JSONDecodeError, jsonschema.exceptions.ValidationError):
            pass
    return result

valid_gt = extract_valid_cookies(ground_truth_json)

# Chưa quan tâm

In [ ]:
# alpaca_prompt = Copied from above
FastLanguageModel.for_inference(model) # Enable native 2x faster inference
inputs = tokenizer(
[alpaca_prompt.format("Continue the fibonnaci sequence.", # instruction
        "1, 1, 2, 3, 5, 8", # input
        "", # output - leave this blank for generation!
    )], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer = text_streamer, max_new_tokens = 128)

<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [ ]:
model.save_pretrained("lora_model")  # Local saving
tokenizer.save_pretrained("lora_model")
# model.push_to_hub("your_name/lora_model", token = "...") # Online saving
# tokenizer.push_to_hub("your_name/lora_model", token = "...") # Online saving

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

In [ ]:
if False:
    from unsloth import FastLanguageModel
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name = "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        max_seq_length = max_seq_length,
        dtype = dtype,
        load_in_4bit = load_in_4bit,
    )
    FastLanguageModel.for_inference(model) # Enable native 2x faster inference

# alpaca_prompt = You MUST copy from above!

inputs = tokenizer(
[
    alpaca_prompt.format(
        "What is a famous tall tower in Paris?", # instruction
        "", # input
        "", # output - leave this blank for generation!
    )
], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer = text_streamer, max_new_tokens = 128)

You can also use Hugging Face's `AutoModelForPeftCausalLM`. Only use this if you do not have `unsloth` installed. It can be hopelessly slow, since `4bit` model downloading is not supported, and Unsloth's **inference is 2x faster**.

In [ ]:
if False:
    # I highly do NOT suggest - use Unsloth if possible
    from peft import AutoPeftModelForCausalLM
    from transformers import AutoTokenizer
    model = AutoPeftModelForCausalLM.from_pretrained(
        "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        load_in_4bit = load_in_4bit,
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")

### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
